Fix NPE in RequestContextSubscriber

robotmrv opened the pull request:
RequestContextSubscriber could cause NPE if Mono/Flux.subscribe
was invoked outside of Web Context.
In addition it replaced source Context with its own without respect
to old data.
Now Request Context Data is Propagated within holder class and
it is added to existing reactor Context

Fixes gh-7228

pivotal-issuemaster commented:
@robotmrv Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7235)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=7235) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@robotmrv Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7235)!

rwinch commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +642,58 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+	public void publisherWhenHooksInitAndSubscribedOutsideOfWebContextThenShouldNotThrowNpe() throws Exception {
+		Mono<Context> publisher = Mono.subscriberContext();
+		this.function.afterPropertiesSet();			// Hooks.onLastOperator() initialized
+//		this.function.destroy();					// Hooks.onLastOperator() released
```
Is there a reason that the tests do not invoke .destroy()?

robotmrv replied:
leave it here just to mark that Hook is not released.  
but maybe it is better to move `this.function.destroy();` to the `cleanup()`

robotmrv replied:
moved `this.function.destroy()` to the `cleanup()` method

robotmrv commented:
I have doubts about  `ServletOAuth2AuthorizedClientExchangeFilterFunction.createRequestContextSubscriber()`  
It creates RequestContextSubscriber on every subscribe()/block() even outside of Security or WebContext and propagates empty DataHolder.  
Maybe it would be better to create RequestContextSubscriber only in case `if (authentication != null || requestAttributes != null)` ?

rwinch commented:
@robotmrv - It looks like something needs to change to me as well. The ThreadLocal should only be invoked inside of defaultRequest which is guaranteed to be invoked on the original Servlet container Thread. I'm going to let @jgrandja take a look at this since he worked on some of the cleanup in this class.

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +643,101 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
 
+	// gh-7228
+	@Test
+	public void publisherWhenHooksInitAndSubscribedOutsideOfWebContextThenShouldNotThrowNpe() throws Exception {
```
If I comment out the following in `createRequestContextSubscriber()`

```
//		if (authentication == null && request == null && response == null) {
//			return delegate;
//		}
```

....the test still passes. I'm not sure this test makes sense?

robotmrv replied:
because this test does not checks this part (publisherWhenHooksInitAndSubscribedOutsideOfWebContextAndOutsideOfSecurityContextAndWithParentContextThenShouldNotReplaceParentContextAndShouldNotAddRequestContext() tests it)   
The check you commented is intended to skip reactor context population with empty RequestContextDataHolder 
 values as it is not useful data (especially outside of web context). This is just optimization ( here my motivation https://github.com/spring-projects/spring-security/pull/7235#issuecomment-520127370 )

robotmrv replied:
NPE is fixed by introducing RequestContextDataHolder which could contain null values,   
It also reduces number of occupied keys in reactor context

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +643,101 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+
+	// gh-7228
+	@Test
+	public void publisherWhenHooksInitAndSubscribedOutsideOfWebContextAndOutsideOfSecurityContextAndWithParentContextThenShouldNotReplaceParentContextAndShouldNotAddRequestContext() throws Exception {
```
I'm not sure about this test and the next one. It's quite low level and not directly testing the API of `ServletOAuth2AuthorizedClientExchangeFilterFunction`. Most tests are directly testing `filter()` and  `defaultRequest()`, whereas these 2 tests are directly using the Reactor API's.

The preference would be to change these tests so they are aligned with the existing `filter()` tests.

robotmrv replied:
The main issue was that current instrumentation affects all application   
So this tests check if it is possible to performe operations on Publishers without web/sucurity context with initialized reactor instrumentation (which is ServletOAuth2AuthorizedClientExchangeFilterFunction functionality)

jgrandja replied:
Both of these tests do not directly test the API of `ServletOAuth2AuthorizedClientExchangeFilterFunction` but instead are testing internal implementation details. In order to simplify the tests, please make `RequestContextSubscriber` package-private along with the constructor. This will allow you to test the constructed state of `RequestContextSubscriber` using `currentContext()`. Please remove these 2 tests after you have added the new ones. Thanks.

robotmrv replied:
as this PR fixes scenario without web/security context (see https://github.com/spring-projects/spring-security/pull/7235#discussion_r316789550 ) and in this scenario `ExchangeFilterFunction` API of `ServletOAuth2AuthorizedClientExchangeFilterFunction` is not used  
but `ServletOAuth2AuthorizedClientExchangeFilterFunction#afterPropertiesSet` is used and it is part of API it is makes sense to test these scenarios without usage of `ExchangeFilterFunction` part of API.  

Other scenarios with direct usage of `ExchangeFilterFunction` API are covered by existing tests. 
I will make `RequestContextSubscriber` package-private but test will not change very much as they could not use `ExchangeFilterFunction` API by scenario.   

These test could be moved to a new test class but I do not see point in this as they still test `ServletOAuth2AuthorizedClientExchangeFilterFunction#afterPropertiesSet` API.

robotmrv commented:
@jgrandja  I've fixed codestyle according to convention standard and removed `publisherWhenHooksInitAndSubscribedOutsideOfWebContextThenShouldNotThrowNpe()`  
Regarding other tests:  
- They check reactor Hook which is part of the ServletOAuth2AuthorizedClientExchangeFilterFunction 
- `RequestContextSubscriber` is low level implementation which should be checked 
- these tests cover scenarios from gh-7228 affect of global reactor hook on whole application - it should not affect unrelated functionality (e.g. reactive DB init on application startup etc.)
- I do not know how to test it better  
so I leave them as is  
If you still have concerns about them please let me know.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunction.java:
```diff
@@ -484,6 +494,9 @@ private ClientRequest bearer(ClientRequest request, OAuth2AuthorizedClient autho
 		}
 		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		if (authentication == null && request == null && response == null) {
+			return delegate;
```
This is misleading since it doesn't create and return a `RequestContextSubscriber` instance, but instead returns the "current" `Subscriber`. I think we should return `null` for these cases and ensure `Hooks.onLastOperator` is NOT called in `afterPropertiesSet()`. Please add a test for this as well.

robotmrv replied:
Hooks.onLastOperator() - applies global lift function on `subscribe()/block()` and it should be better installed once per application life circle.  
As Lift function is invoked only at the moment of `subscribe()` lazily (not at the moment of application context / Filter initialization) it is not possible to unsure is there web/security context or not.  

Not sure we are on the same page with case I try to fix  
One of the cases:  
I want to initialize application with some data. I have my reactive service API to do it. 
I make `ApplicationRunner` or just Application listener which invokes my reactive API on startup and call subscribe()/block(). 
So If my API implementation uses reactor context (e.g. reactive transactions, or just some custom implementation) with initialized hook from filter and without current fix I will get NPE, because `RequestContextSubscriber` will populate reactor context with `null` values (as there is no Web and Security Context) and this is forbidden.  

Another scenario - my reactive API is invoked form RabbitMQ listener and there is no web context - as the result NPE.  

So what PR fixes:  
1. it ensures that no `null` values are propagated to the reactor context.  
2. As reactor context recommended to be as small as possible (see javadocs) this fix reduces number of used entries from 4 to 1/0.  
3. `RequestContextSubscriber` replaced original context so data from other hooks could be lost. This fix ensures that `RequestContextSubscriber` will not replace all parent context but adds values to it.  
4. In described case `RequestContextDataHolder` will be created with `null` values and will be useless as it is not required at all for the operation (application initialization), so it is make sense not to populate reactor context with useless data because of p.2 and not to create additional `Subscriber` to reduce allocation and invocation stack size on each emit or `currentContext()`

Lift function can't return `null` as later we will get NPE so `delegate` subscriber looks fine to me as it means NOOP lift function.   

@jgrandja, do you still want to return `null` from lift function and create `RequestContextSubscriber` in any case with empty `RequestContextDataHolder`?

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -146,9 +153,10 @@ public void setup() {
 
 	@After
-	public void cleanup() {
+	public void cleanup() throws Exception {
```
Please revert the 2 changes in `cleanup()` as I believe it was added as a result of the removed test.

robotmrv replied:
No.  
It removes reactor Hook from the filter,  i.e. makes cleanup of the global state.   
So it should be better here.

robotmrv commented:
@jgrandja , 
I've updated tests, please check

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +641,69 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
 
+	// gh-7228
+	@Test
+	public void afterPropertiesSetWhenHooksInitAndPublisherSubscribedOutsideOfWebContextAndOutsideOfSecurityContextThenSubscriptionShouldNotThrowException() throws Exception {
```
Let's shorten this name to `afterPropertiesSetWhenHooksInitAndOutsideWebSecurityContextThenShouldNotThrowException`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +641,69 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+	}
+
+	@Test
+	public void createRequestContextSubscriberIfNecessaryWhenOutsideOfWebContextAndOutsideOfSecurityContextThenShouldReturnOriginalSubscriber() throws Exception {
```
Please rename to `createRequestContextSubscriberIfNecessaryWhenOutsideWebSecurityContextThenReturnOriginalSubscriber`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +641,69 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+
+	// gh-7228
+	@Test
+	public void requestSubscriberWhenHasRequestAndHasResponseThenShouldAddToParentContextRequestContextDataHolder() throws Exception {
```
Please rename to `createRequestContextSubscriberWhenRequestResponseProvidedThenCreateWithParentContext`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +641,69 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+
+	// gh-7228
+	@Test
+	public void requestSubscriberWhenHasAuthenticationThenShouldAddToParentContextRequestContextDataHolder() throws Exception {
```
Please rename to `createRequestContextSubscriberWhenAuthenticationProvidedThenCreateWithParentContext`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -635,6 +641,69 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
+	public void requestSubscriberWhenHasAuthenticationThenShouldAddToParentContextRequestContextDataHolder() throws Exception {
+		testRequestContextSubscriber(null, null, this.authentication);
+	}
+
```
Can you also add a test where `RequestContextSubscriber` constructor uses `parentContext`

```java
			if (parentContext.hasKey(REQUEST_CONTEXT_DATA_HOLDER)) {
				context = parentContext;

``

robotmrv replied:
@jgrandja 
I've added test and renamed existing as you suggested.  
please check

robotmrv commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java:
```diff
@@ -643,32 +643,53 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
 
+	@Test
+	public void createRequestContextSubscriberWhenParentContextHasDataHolderThenShouldReuseParentContext() throws Exception {
+		RequestContextDataHolder testValue = mock(RequestContextDataHolder.class);
```
@jgrandja done

jgrandja commented:
Thank you for all the updates @robotmrv. This is now in master.
